File: pypp_cli/test_dir/cpp/src/exceptions/custom_exceptions.cpp
#include "exceptions/custom_exceptions.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include <string>

namespace me {
class _PrivateCustomException : public pypp::PyppException {
  public:
    explicit _PrivateCustomException(const pypp::PyStr &msg)
        : pypp::PyppException(pypp::PyStr("_PrivateCustomException: ") + msg) {}
};

void custom_exception_fn() {
    pypp::print(pypp::PyStr("pypp CUSTOM EXCEPTION RESULTS:"));
    try {
        throw CustomException(
            pypp::PyStr("This is a custom exception message."));
    } catch (const CustomException &pypp_e) {
        std::string e = pypp_e.what();
        pypp::print(pypp::PyStr("custom exception caught: ") +
                    pypp::to_pystr(e));
    }
    try {
        throw ChildException(pypp::PyStr("This is a child exception message."));
    } catch (const ChildException &pypp_e) {
        std::string e = pypp_e.what();
        pypp::print(pypp::PyStr("child exception caught: ") +
                    pypp::to_pystr(e));
    }
}

} // namespace me

File: pypp_cli/test_dir/python/src/exceptions/custom_exceptions.py
from multiprocessing.managers import ValueProxy
from pypp_python import exception


@exception
class CustomException(Exception):
    pass


@exception
class ChildException(CustomException):
    pass


@exception
class _PrivateCustomException(Exception):
    pass


@exception
class CustomValueError(ValueError):
    pass


def custom_exception_fn():
    print("pypp CUSTOM EXCEPTION RESULTS:")
    try:
        raise CustomException("This is a custom exception message.")
    except CustomException as e:
        print("custom exception caught: " + str(e))
    try:
        raise ChildException("This is a child exception message.")
    except ChildException as e:
        print("child exception caught: " + str(e))
    try:
        raise CustomValueError("This is a custom value error message.")
    except ValueError as e:
        print("custom value error caught: " + str(e))
